### ISSUE TITLE
Request::getBodyParams() and the treatment of Content-Type

### ISSUE THREAD

**Issue by PowerGamer1** [yiisoft/yii2#14522]
The current implementation of `Request::getBodyParams()` was (badly) designed to behave as follows:

1. If the client passed `Content-Type` configured in `Request::parsers` or there is fallback '*' parser - parse it accordingly.
2. If the client passed `Content-Type` NOT configured in `Request::parsers` (and there is no fallback '*' parser) pretend its "application/x-www-form-urlencoded" and parse it accordingly -- at least whoever wrote `Request::getBodyParams()` wanted it to work like this but in reality it DOES NOT in all cases (see below).
3. If the client did not pass a `Content-Type` pretend its "application/x-www-form-urlencoded" and parse it accordingly.

Lets see what HTTP 1.1 spec has to say about this:
https://tools.ietf.org/html/rfc7231#section-3.1.1.5
```
A sender that generates a message containing a payload body SHOULD
   generate a Content-Type header field in that message unless the
   intended media type of the enclosed representation is unknown to the
   sender.  If a Content-Type header field is not present, the recipient
   MAY either assume a media type of "application/octet-stream"
   ([RFC2046], Section 4.5.1) or examine the data to determine its type.
```

https://tools.ietf.org/html/rfc7231#section-6.5.13
```
6.5.13.  415 Unsupported Media Type
   The 415 (Unsupported Media Type) status code indicates that the
   origin server is refusing to service the request because the payload
   is in a format not supported by this method on the target resource.
   The format problem might be due to the request's indicated
   Content-Type or Content-Encoding, or as a result of inspecting the
   data directly.
```

Lets see what PHP does when a client sends POST request with a data encoded in "application/x-www-form-urlencoded" or "multipart/form-data":
1. When the Content-Type is set to "application/x-www-form-urlencoded" or "multipart/form-data" PHP decodes the data and fills $_POST array.
2. When the Content-Type is absent or is not set to "application/x-www-form-urlencoded" or "multipart/form-data" PHP does NOT try to process the data from request body and the $_POST array remains empty.

Point 2. of PHP behavior above explains why the point 2. of intended behavior of `Request::parsers` algorithm will not work as intended:
```PHP
            } elseif ($this->getMethod() === 'POST') {
                // PHP has already parsed the body so we have all params in $_POST
                $this->_bodyParams = $_POST;
            }
```

Considering all of the above `Request::getBodyParams()` should be rewritten to function as follows:
1. If the client passed `Content-Type` configured in `Request::parsers` or there is fallback '*' parser - parse it accordingly. - No change.
2. If the client passed `Content-Type` NOT configured in `Request::parsers` (and there is no fallback '*' parser) - don't break PHP conformant behavior and follow HTTP 1.1 spec guidelines - refuse to serve the response with "415 Unsupported Media Type" status code.
3. If the client did not pass a `Content-Type` do the same as point 2.

By default, `Request` should come with two built-in parsers configured - which will handle "Content-Type: application/x-www-form-urlencoded" and "Content-Type: multipart/form-data" for all request methods with body (the other Content-Type values with get error 415).

**Comment by PowerGamer1** [comment]
Also, another problem with `Request::getBodyParams()` - it ignores the return value of `mb_parse_str()` call which may be unsuccessful.

**Comment by cebe** [comment]
> If the client passed Content-Type NOT configured in Request::parsers (and there is no fallback '*' parser) - don't break PHP conformant behavior and follow HTTP 1.1 spec guidelines - refuse to serve the response with "415 Unsupported Media Type" status code.

does PHP return a "415 Unsupported Media Type" in that case?
If not, by taking `$_POST` from PHP we are exactly following PHP behavior, i.e. not trying to parse the content.

The use of `mb_parse_str` is imo what the RFC states as "or examine the data to determine its type." If it succeeds, it was a regular post request, if not, we leave `$this->_bodyParams` empty as PHP does.
We could change that to return 415 though.

**Comment by PowerGamer1** [comment]
> does PHP return a "415 Unsupported Media Type" in that case?

PHP of course does not return 415 error as it never returns any HTTP errors to the client by itself - its the responsibility of the PHP script. In case of Yii2 framework and `Request` class in particular I would expect it (i.e. returning 415 code) to be the responsibility of `Request` class.

>  If not, by taking  $_POST  from PHP we are exactly following PHP behavior, i.e. not trying to parse the content.

I think the intent of `getBodyParams()` code was different:
```PHP
            } elseif ($this->getMethod() === 'POST') {
                // PHP has already parsed the body so we have all params in $_POST
                $this->_bodyParams = $_POST;
```
The comment in the code above suggests that code expected PHP to parse anything regardless of Content-Type. Also, "following PHP behavior" should mean "be standard compliant as PHP" (i.e. do not try to parse unrecognized Content-Type).

> The use of  mb_parse_str  is imo what the RFC states as "or examine the data to determine its type."

Nope, "examine the data" implies you try to recognize its media-type but `mb_parse_str` just assumes the data is in "application/x-www-form-urlencoded" format and parses it accordingly. `mb_parse_str` does not even recognize "multipart/form-data" - it parses it as if it was "application/x-www-form-urlencoded" - successfully but completely incorrectly. Also, the HTTP spec basically discourages the examination of the data to determine its type (since its never 100% reliable) in the next paragraph:

```
   In practice, resource owners do not always properly configure their
   origin server to provide the correct Content-Type for a given
   representation, with the result that some clients will examine a
   payload's content and override the specified type.  Clients that do
   so risk drawing incorrect conclusions, which might expose additional
   security risks (e.g., "privilege escalation").  Furthermore, it is
   impossible to determine the sender's intent by examining the data
   format: many data formats match multiple media types that differ only
   in processing semantics.  Implementers are encouraged to provide a
   means of disabling such "content sniffing" when it is used.
```

So yes, I think returning 415 when appropriate would be the best.

**Comment by klimov-paul** [comment]
> If the client did not pass a Content-Type do the same as point 2.

This statement contradicts your own RFC references:

> If a Content-Type header field is not present, the recipient
   MAY either assume a media type of "application/octet-stream"

This means if no 'Content-Type' is specified application may assume it is 'application/octet-stream' and parse it in whatever it consider correct way, which includes assumption is is 'form-urlencoded' data.

**Comment by PowerGamer1** [comment]
```TEXT
If a Content-Type header field is not present, the recipient
MAY either assume a media type of "application/octet-stream"
([RFC2046], Section 4.5.1) or examine the data to determine its type.
```
1. Recipient can `assume a media type of "application/octet-stream"`. This does not mean if it assumes so it must be able to somehow understand and handle it. In particular, if `Request::parsers` has no "application/octet-stream" and no fallback '*' parser configured, it should return 415 according to my algorithm.
2. Recipient can "examine the data to determine its type" - practically impossible to do reliably and shouldn't be attempted.

So nothing in my algorithm contradicts RFC.

> This means if no 'Content-Type' is specified application may assume it is 'application/octet-stream' **and parse it in whatever it consider correct way, which includes assumption is is 'form-urlencoded' data.**

I don't see how you deduced the part in bold from what RFC says. The "application/octet-stream" and "form-urlencoded" are completely different content types, you can't guarantee that "application/octet-stream" binary data will always be correctly parsable as "form-urlencoded".

**Comment by klimov-paul** [comment]
> I don't see how you deduced the part in bold from what RFC says

I am unable to explain it any better: no 'content-type' is equal (can be assumed equal to be correct) to 'application/octet-stream' according to RFC.
'application/octet-stream' content type is a generic one, means there is no strict rule on how it should be processed. Thus I may attempt to parse it in the same way as 'form-urlencoded' - it may fail, but it does not mean I can not attempt to do so. Such behavior does not contradict RFC anyhow.

Throwing 415 HTTP error on empty 'Content-type' seems to be risk prone as some 'lazy' client may not explicitly set it. For instance

**Comment by klimov-paul** [comment]
Actually you may achieve comlete compliance with RFC using current 2.0.x code.
You just need to create parser for 'form/urnencoded' and setup it to `Request::$parsers` by 'form/urnencoded' key. Then create a dummy parser, which will throw `HttpException` with 415 code on its `parse()` incovation, and setup it as a fallback parser with key '*'.
After that your application will accept only 'form/urnencoded' data and return 415 error in any other case.

Our treatment of request body as 'form/urnencoded', which you complain on, is just a default fallback parser, which you may override without any problem.

**Comment by PowerGamer1** [comment]
> 'application/octet-stream' content type is a generic one, means there is no strict rule on how it should be processed. Thus I may attempt to parse it in the same way as 'form-urlencoded' - it may fail, but it does not mean I can not attempt to do so. Such behavior does not contradict RFC anyhow.

'application/octet-stream' - is just that - some binary data you can't do much with other then store or transfer it somewhere. Attempting to guess and interpret a more specific data from 'application/octet-stream' does not contradict the RFC, yes, it just does not make sense - why would you do something like that at all?

> Throwing 415 HTTP error on empty 'Content-type' seems to be risk prone as some 'lazy' client may not explicitly set it.

For such "lazy" client PHP won't parse the body so why should Yii2 have to try to parse it? Don't see any risks here.

> Actually you may achieve complete compliance with RFC using current 2.0.x code.

Framework should not require obscure manual setup to become RFC-compliant. It must be compliant out-of-the-box.

> Our treatment of request body as 'form/urnencoded', which you complain on, is just a default fallback parser, which you may override without any problem.

Again, you can't 100% guarantee that even IF you successfully parse unknown request body data as 'form/urnencoded' it was actually representing a 'form/urnencoded' data. If the 'Content-Type' is unsupported by the server or not specified in the request the smart thing to do would be to rise an error rather than risk parsing data incorrectly. In particular PHP itself does NOT try to parse a body of such requests - that is a smart behavior Yii2 should follow too.

And at the very least the code I quoted in my OP:
```PHP
            } elseif ($this->getMethod() === 'POST') {
                // PHP has already parsed the body so we have all params in $_POST
                $this->_bodyParams = $_POST;
            }
```
is plain wrong as the assumption expressed in the comment is incorrect. I explained that in my OP too.

**Comment by klimov-paul** [comment]
Resolved by commit e7943971d25c3021a47deef2fa377ec2d80f8169
